free/src/main/scala/cats/free/FreeT.scala

```diff
@@ -0,0 +1,263 @@
+    flatMap(a => pure(f(a)))
+
+  /** Binds the given continuation to the result of this computation. */
+  final def flatMap[B](f: A => FreeT[S, M, B]): FreeT[S, M, B] =
+    FlatMapped(this, f)
+
+  /**
+   * Changes the underlying `Monad` for this `FreeT`, ie.
+   * turning this `FreeT[S, M, A]` into a `FreeT[S, N, A]`.
+   */
+  def hoist[N[_]](mn: M ~> N): FreeT[S, N, A] =
+    step match {
+      case e @ FlatMapped(_, _) =>
+        FlatMapped(e.a.hoist(mn), e.f.andThen(_.hoist(mn)))
+      case Suspend(m) =>
+        Suspend(mn(m))
+    }
+
+  /** Change the base functor `S` for a `FreeT` action. */
+  def interpret[T[_]](st: S ~> T)(implicit M: Functor[M]): FreeT[T, M, A] =
+    step match {
+      case e @ FlatMapped(_, _) =>
+        FlatMapped(e.a.interpret(st), e.f.andThen(_.interpret(st)))
+      case Suspend(m) =>
+        Suspend(M.map(m)(_.map(s => st(s))))
+    }
+
+  /**
+   * Runs to completion, mapping the suspension with the given transformation
+   * at each step and accumulating into the monad `M`.
+   */
+  def foldMap(f: S ~> M)(implicit MR: Monad[M], RT: RecursiveTailRecM[M]): M[A] = {
+    def go(ft: FreeT[S, M, A]): M[FreeT[S, M, A] Xor A] =
+      ft match {
+        case Suspend(ma) => MR.flatMap(ma) {
+          case Xor.Left(a) => MR.pure(Xor.Right(a))
+          case Xor.Right(sa) => MR.map(f(sa))(Xor.right)
+        }
+        case g @ FlatMapped(_, _) => g.a match {
+          case Suspend(mx) => MR.flatMap(mx) {
+            case Xor.Left(x) => MR.pure(Xor.left(g.f(x)))
+            case Xor.Right(sx) => MR.map(f(sx))(g.f andThen Xor.left)
+          }
+          case g0 @ FlatMapped(_, _) => MR.pure(Xor.left(g0.a.flatMap(g0.f(_).flatMap(g.f))))
+        }
+      }
+
+    MR.tailRecM(this)(go)
+  }
+
+  /** Evaluates a single layer of the free monad */
+  def resume(implicit S: Functor[S], MR: Monad[M], RT: RecursiveTailRecM[M]): M[A Xor S[FreeT[S, M, A]]] = {
+    def go(ft: FreeT[S, M, A]): M[FreeT[S, M, A] Xor (A Xor S[FreeT[S, M, A]])] =
+      ft match {
+        case Suspend(f) => MR.map(f)(as => Xor.right(as.map(S.map(_)(pure(_)))))
+        case g1 @ FlatMapped(_, _) => g1.a match {
+          case Suspend(m1) => MR.map(m1) {
+            case Xor.Left(a) => Xor.left(g1.f(a))
+            case Xor.Right(fc) => Xor.right(Xor.right(S.map(fc)(g1.f(_))))
+          }
+          case g2 @ FlatMapped(_, _) => MR.pure(Xor.left(g2.a.flatMap(g2.f(_).flatMap(g1.f))))
+        }
+      }
+
+    MR.tailRecM(this)(go)
+  }
+
+  /**
+   * Runs to completion, using a function that maps the resumption from `S` to a monad `M`.
+   */
+  def runM(interp: S[FreeT[S, M, A]] => M[FreeT[S, M, A]])(implicit S: Functor[S], MR: Monad[M], RT: RecursiveTailRecM[M]): M[A] = {
+    def runM2(ft: FreeT[S, M, A]): M[FreeT[S, M, A] Xor A] =
+      MR.flatMap(ft.resume) {
+        case Xor.Left(a) => MR.pure(Xor.right(a))
+        case Xor.Right(fc) => MR.map(interp(fc))(Xor.left)
+      }
+    MR.tailRecM(this)(runM2)
```

Comment:
you should do: `RT.sameType(MR).tailRecM(this)(runM2)` so we have the compiler check we have the right type on `RecursiveTailRecM`